### ISSUE TITLE
Introduce tag for use of adaptive meshes

### ISSUE THREAD

**Issue by fsimonis** [precice/precice#1293]
Introducing re-meshing leads to additional synchronisation points and re-paritioning work.

To reduce this cost as much as possible, I suggest requiring adaptive meshes be tagged as such.

```xml
<mesh name="Mesh">
  ...
</mesh>

<participant name="Sender">
  <use-mesh name="Mesh" provide="true" adaptive="true" />
</participant>

<participant name="Receiver">
  <use-mesh name="Mesh" from="Sender" />
</participant>
```

This allows us to:
* disable all re-meshing features by default.
  Existing configurations will work out-of-the-box without any performance impact.
* enforce that only Sender may use `resetMesh(meshID)` to reset and redefine the mesh.
  This restricts which meshes need repartitioning on which participants.
* enable the intra-participant handshake only for Sender, as it provides an adaptive mesh.
  The Receiver does not need this handshake and can skip it as it will never be able to call `resetMesh` itself.
* enable re-initialization only between Sender and Receiver (everyone using Mesh from Sender)
  Other participants will never re-initialize as they don't come in contact with participants providing adaptive meshes.

**Comment by uekerman** [comment]
I like. Just thinking a bit about the name. As a mathematical term, "adaptive" is sth different. ˋdynamicˋ could be another option or ˋremeshingˋ or ˋallow-remeshingˋ?

**Comment by fsimonis** [comment]
Further alternatives:
* a reverse attribute such as `static=true` or
* a kind attribute `kind="static" `, `kind="dynamic" `
* use a namespace `use-mesh:static`, `use-mesh:dynamic`

**Comment by uekerman** [comment]
I like `static="true"` as well.

**Comment by fsimonis** [comment]
@uekerman and I converged towards the attribute `dynamic="true"`.

**Comment by fsimonis** [comment]
For the current implementation goal we need a global switch, so ˋ<precice-configuration allow-remeshing="True" />ˋ sounds like a good fit.

**Comment by MakisH** [comment]
In preCICE v3 format, the original config would be:

```xml
<mesh name="Mesh" dimensions="3">
  ...
</mesh>

<participant name="Sender">
  <provide-mesh name="Mesh" dynamic="true" />
</participant>

<participant name="Receiver">
  <receive-mesh name="Mesh" from="Sender" />
</participant>
```

As a user, I would expect the static/dynamic to be associated with the mesh, not with the participant:

```xml
<mesh name="Mesh" dimensions="3" dynamic=`true`>
  ...
</mesh>
```

But then I would expect the `resetMesh()` to only be allowed for a `provide-mesh`.

**Comment by MakisH** [comment]
One could even make this a separate tag. As a participant can send/receive mesh, a participant could also redefine meshes:

```xml
<participant name="Sender">
  <provide-mesh name="Mesh" />
  <redefine-mesh name="Mesh" allowed="true" />
</participant>
```

**Comment by fsimonis** [comment]
> But then I would expect the resetMesh() to only be allowed for a provide-mesh.

Thats an interesting observation.

The mesh can only be directly changed/set by the participant that provides it.
So both locations are equivalent.

What is easier to follow from a user-standpoint:

* The mesh M may be remeshed. So only the participant that provides the mesh may remesh and all participants that use it may need to occasionally do extra work.
* The participant P provides the mesh M and may remesh it. So other participants using mesh M may need to do some extra work now and then.

**Comment by MakisH** [comment]
> Given that the remeshing functionality must be implemented and activated in the adapter, it may be easier to tell the user to enable remeshing in the participant.

Remember that a participant can generally be "implemented" by different solvers. So, connecting the participant to the adapter feature does not necessarily help.

Situations to compare to would be the nearest-project mapping and the direct mesh access (both indeed defined at the participant level). I can understand a consistency argument, then.

Another situation is the geometric multiscale: as a mapping, it is defined at the participant level, but the dimensions remain a property of the mesh, so each participant that provides or receives a (in the future) 1D mesh, should be able to handle 1D meshes.

> There is barely a technical reason to do the one over the other (adding it to the mesh is slightly easier to handle). So I would prioritize the user-experience and documentability.

I agree that there is little reason to prefer any of the two. I just associate the ability to change with the mesh itself (because the mesh is what changes).

If one participant changes the mesh, how/when do other participants find out? Could the one or the other way prevent some duplication and/or erroneous state?

**Comment by fsimonis** [comment]
> If one participant changes the mesh, how/when do other participants find out? Could the one or the other way prevent some duplication and/or erroneous state?

preCICE needs to deduce which participants need to handle remeshing at configuration time. Handshakes between participants are required as soon as remeshing is allowed.

**Comment by MakisH** [comment]
Then I guess it would be like:

- Option 1 (mesh): Mesh was updated -> Identify all participants that use this mesh -> Trigger handshaking between these participants (and intra-participant)
- Option 2 (participant): Participant updated the mesh -> Trigger handshaking between all participants that use this mesh (and intra-participant)

Associating this with the participant would skip the identifying which participants use this mesh, which is probably trivial anyway.

**Comment by fsimonis** [comment]
Implementation is not subject of this discussion.

The only important requirement is that we need information which mesh allows remeshing in the configuration, as deducing remesh requests at runtime results in an expensive performance hit for everyone who doesn't use remeshing.

TLDR; The question is where to configure it.

**Comment by fsimonis** [comment]
I'll unassign myself as this is currently an unnecessary feature.
